lib/trino-parquet/src/main/java/io/trino/parquet/AbstractParquetDataSource.java

```diff
@@ -209,9 +211,24 @@ private <K> ListMultimap<K, ChunkReader> readLargeDiskRanges(ListMultimap<K, Dis
             return ImmutableListMultimap.of();
         }
 
+        int maxBufferSizeBytes = toIntExact(options.getMaxBufferSize().toBytes());
+        checkArgument(maxBufferSizeBytes > 0, "maxBufferSize must by larger than zero but is %s bytes", maxBufferSizeBytes);
         ImmutableListMultimap.Builder<K, ChunkReader> slices = ImmutableListMultimap.builder();
         for (Map.Entry<K, DiskRange> entry : diskRanges.entries()) {
-            slices.put(entry.getKey(), new ReferenceCountedReader(entry.getValue()));
+            DiskRange range = entry.getValue();
+            long endOffset = range.getOffset() + range.getLength();
+            long offset = range.getOffset();
+            while (offset + maxBufferSizeBytes < endOffset) {
+                DiskRange chunkRange = new DiskRange(offset, maxBufferSizeBytes);
+                slices.put(entry.getKey(), new ReferenceCountedReader(chunkRange));
+                offset += maxBufferSizeBytes;
+            }
+
+            long lengthLeft = endOffset - offset;
+            if (lengthLeft > 0) {
+                DiskRange chunkRange = new DiskRange(offset, toIntExact(lengthLeft));
+                slices.put(entry.getKey(), new ReferenceCountedReader(chunkRange));
```

Comment:
This logic doesn't belong here. This method is supposed to read the ranges as provided in the `diskRanges` argument. The splitting between small ranges and large ranges, and the logic of how to do that split belongs in the caller.

Reply:
well, it is already merging `diskRanges` in `io.trino.parquet.AbstractParquetDataSource#mergeAdjacentDiskRanges`.
How is splitting different?